geode-core/src/main/java/org/apache/geode/internal/net/ByteBufferSharingImpl.java

```diff
@@ -0,0 +1,91 @@
+ * the License.
+ */
+
+package org.apache.geode.internal.net;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An {@link AutoCloseable} meant to be acquired in a try-with-resources statement. The resource (a
+ * {@link ByteBuffer}) is available (for reading and modification) in the scope of the
+ * try-with-resources. Collaborates with {@link ByteBufferReferencing} to ensure final dereference
+ * returns the {@link ByteBuffer} to the pool.
+ */
+class ByteBufferSharingImpl implements ByteBufferSharing {
+
+  private final ByteBuffer buffer;
+  private final Lock lock;
+  private final ByteBufferReferencing referencing;
+
+  /**
+   * This constructor is for use only by the owner of the shared resource (a {@link ByteBuffer}).
+   * Furthermore, this constructor is only for use in constructing the very first {@link
+   * ByteBufferSharingImpl} for the resource. Subsequent instances must be constructed via the copy
+   * constructor.
+   *
+   * A resource owner calls this constructor and retains the reference and never hands it out.
+   * Instead, the owner hands out copies, constructed via the copy constructor.
+   *
+   * This constructor acquires no lock and does not modify the reference count through the {@link
+   * ByteBufferReferencing} (passed in here.)
+   */
+  ByteBufferSharingImpl(final ByteBuffer buffer,
+      final ByteBufferReferencing referencing) {
+    this(buffer, new ReentrantLock(), referencing);
+  }
+
+  /**
+   * This constructor is for use only by the owner of the shared resource. It's used for handing out
+   * references to the shared resource. So it does reference counting and also acquires a lock.
+   *
+   * Resource owners will typically return the result of this constructor, to a caller. That caller
+   * binds that reference to a variable in a try-with-resources statement and relies on the
+   * AutoCloseable protocol to invoke close() on the object at the end of the block.
+   */
+  ByteBufferSharingImpl(final ByteBufferSharingImpl other) {
+    this(other.buffer, other.lock, other.referencing);
+    referencing.addReference();
+    lock.lock();
+  }
+
+  private ByteBufferSharingImpl(final ByteBuffer buffer,
+      final Lock lock,
+      final ByteBufferReferencing referencing) {
+    this.buffer = buffer;
+    this.lock = lock;
+    this.referencing = referencing;
+  }
+
+  @Override
+  public ByteBuffer getBuffer() {
+    return buffer;
+  }
+
+  @Override
+  public void close() {
+    final int usages = referencing.dropReference();
+    if (usages > 0) {
+      /*
+       * We don't unlock when the very last usage goes away. The resource owner holds the first
+       * usage and there is no lock associated with that one. Subsequent usages have a lock and
+       * so require a corresponding unlock.
+       */
+      lock.unlock();
```

Comment:
I think this introduces a bug in the case when "close" finds the reference still in use. In that case "close" will call unlock but that will fail because ReentrantLock requires that unlock be called by the same thread that called lock (I think). In general it wold be better if the thread that called lock at the beginning of the "try-with-resource" was also the caller of "unlock"

Reply:
Talked to @dschneider-pivotal and we agreed this is not an issue.